### PR TITLE
infer non-empty-string on substr() comparison with constant string

### PR DESCRIPTION
**Pull request by staabm**
refs https://github.com/phpstan/phpstan/issues/6792

### CODE REVIEW

**Comment by staabm** [comment]
memory related errors are also present on latest 1.6.x branch, therefore feel unrelated

**Comment by staabm** [comment]
Todo: verify this doesnt hit the same problem as https://github.com/phpstan/phpstan-src/pull/1275

**Comment by staabm** [comment]
> Todo: verify this doesnt hit the same problem as #1275

done. this one is good to go.

File: tests/PHPStan/Analyser/data/non-empty-string-substr-specifying.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+			assertType('non-empty-string', $s);
+		}
+		assertType('string', $s);
+
+		if (substr($s, $offset) === 'hallo') {
+			assertType('non-empty-string', $s);
+		}
+		assertType('string', $s);
+
+		if (substr($s, $offset, $length) === 'hallo') {
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/non-empty-string-substr-specifying.php:43]: There should be a test for yoda condition (reversed) too.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Analyser/data/non-empty-string-substr-specifying.php (hunk 1 above)
Comment (staabm) [reply tests/PHPStan/Analyser/data/non-empty-string-substr-specifying.php:43]: done 👍

**Comment by staabm** [comment]
Should this be re-targeted for 1.7.x?

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thanks :) Even if there isn't another 1.6.x release, I'm gonna merge it to 1.7.x right away.

**Pull request merged**